File: Microsoft.Recognizers.Text.Number/Extractors/BaseNumberExtractor.cs
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Text.RegularExpressions;
using Microsoft.Recognizers.Resources;

namespace Microsoft.Recognizers.Text.Number
{
    public abstract class BaseNumberExtractor : IExtractor
    {
        internal abstract ImmutableDictionary<Regex, string> Regexes { get; }

        protected virtual string ExtractType { get; } = "";

        public virtual List<ExtractResult> Extract(string source)
        {
            if (string.IsNullOrEmpty(source))
            {
                return new List<ExtractResult>();
            }

            var result = new List<ExtractResult>();
            var matchSource = new Dictionary<Match, string>();
            var matched = new bool[source.Length];
            for (var i = 0; i < source.Length; i++)
            {
                matched[i] = false;
            }

            var collections = Regexes.ToDictionary(o => o.Key.Matches(source), p => p.Value);
            foreach (var collection in collections)
            {
                foreach (Match m in collection.Key)
                {
                    for (var j = 0; j < m.Length; j++)
                    {
                        matched[m.Index + j] = true;
                    }

                    //Keep Source Data for extra information
                    matchSource.Add(m, collection.Value);
                }
            }

            var last = -1;
            for (var i = 0; i < source.Length; i++)
            {
                if (matched[i])
                {
                    if (i + 1 == source.Length || !matched[i + 1])
                    {
                        var start = last + 1;
                        var length = i - last;
                        var substr = source.Substring(start, length);

                        if (matchSource.Keys.Any(o => o.Index == start && o.Length == length))
                        {
                            var srcMatch = matchSource.Keys.First(o => o.Index == start && o.Length == length);
                            var er = new ExtractResult
                            {
                                Start = start,
                                Length = length,
                                Text = substr,
                                Type = ExtractType,
                                Data = matchSource.ContainsKey(srcMatch) ? matchSource[srcMatch] : null
                            };
                            result.Add(er);
                        }
                    }
                }
                else
                {
                    last = i;
                }
            }

            return result;
        }

        protected Regex GenerateLongFormatNumberRegexes(LongFormatType type, string placeholder = CommonNumeric.PlaceHolderDefault)
        {
            Regex addedRegex = null;
            string integerTemplate = CommonNumeric.IntegerTemplateRegex + $@"(?={placeholder})";
            string doubleTemplate = CommonNumeric.DoubleTemplateRegex + $@"(?={placeholder})";
            switch (type)
            {
                case LongFormatType.IntegerNumComma:
                    addedRegex = new Regex(string.Format(integerTemplate, ","), RegexOptions.IgnoreCase | RegexOptions.Singleline);
                    break;
                case LongFormatType.IntegerNumDot:
                    addedRegex = new Regex(string.Format(integerTemplate, Regex.Escape(".")), RegexOptions.IgnoreCase | RegexOptions.Singleline);
                    break;
                case LongFormatType.IntegerNumBlank:
                    addedRegex = new Regex(string.Format(integerTemplate, " "), RegexOptions.IgnoreCase | RegexOptions.Singleline);
                    break;
                case LongFormatType.IntegerNumQuote:
                    addedRegex = new Regex(string.Format(integerTemplate, Regex.Escape("'")), RegexOptions.IgnoreCase | RegexOptions.Singleline);
                    break;
                case LongFormatType.DoubleNumCommaDot:
                    addedRegex = new Regex(string.Format(doubleTemplate, ",", Regex.Escape(".")), RegexOptions.IgnoreCase | RegexOptions.Singleline);
                    break;
                case LongFormatType.DoubleNumDotComma:
                    addedRegex = new Regex(string.Format(doubleTemplate, Regex.Escape("."), ","), RegexOptions.IgnoreCase | RegexOptions.Singleline);
                    break;
                case LongFormatType.DoubleNumBlankComma:
                    addedRegex = new Regex(string.Format(doubleTemplate, " ", ","), RegexOptions.IgnoreCase | RegexOptions.Singleline);
                    break;
                case LongFormatType.DoubleNumBlankDot:
                    addedRegex = new Regex(string.Format(doubleTemplate, " ", Regex.Escape(".")), RegexOptions.IgnoreCase | RegexOptions.Singleline);
                    break;
                case LongFormatType.DoubleNumCommaCdot:
                    addedRegex = new Regex(string.Format(doubleTemplate, ",", "·"), RegexOptions.IgnoreCase | RegexOptions.Singleline);
                    break;
                case LongFormatType.DoubleNumQuoteComma:
                    addedRegex = new Regex(string.Format(doubleTemplate, Regex.Escape("'"), ","), RegexOptions.IgnoreCase | RegexOptions.Singleline);
                    break;
            }
            return addedRegex;
        }
    }

    public enum NumberMode
    {
        //Default is for unit and datetime
        Default,
        //Add 67.5 billion & million support.
        Currency,
        //Don't extract number from cases like 16ml
        PureNumber
    }

    public enum LongFormatType
    {
        // Reference : https://www.wikiwand.com/en/Decimal_mark

        // Value : 1234567.89
        // 1,234,567
        IntegerNumComma,
        // 1.234.567
        IntegerNumDot,
        // 1 234 567
        IntegerNumBlank,
        // 1'234'567
        IntegerNumQuote,
        // 1,234,567.89
        DoubleNumCommaDot,
        // 1,234,567·89
        DoubleNumCommaCdot,
        // 1 234 567,89
        DoubleNumBlankComma,
        // 1 234 567.89
        DoubleNumBlankDot,
        // 1.234.567,89
        DoubleNumDotComma,
        // 1'234'567,89
        DoubleNumQuoteComma
    }
}

File: Java/libraries/recognizers-text-number/src/main/java/com/microsoft/recognizers/text/number/extractors/BaseNumberExtractor.java
package com.microsoft.recognizers.text.number.extractors;

import com.microsoft.recognizers.text.ExtractResult;
import com.microsoft.recognizers.text.IExtractor;
import com.microsoft.recognizers.text.number.LongFormatType;
import com.microsoft.recognizers.text.number.NumberOptions;
import com.microsoft.recognizers.text.number.resources.BaseNumbers;
import com.microsoft.recognizers.text.utilities.Match;
import com.microsoft.recognizers.text.utilities.RegExpUtility;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public abstract class BaseNumberExtractor implements IExtractor {

    protected abstract Map<Pattern, String> getRegexes();

    protected Map<Pattern, Pattern> getAmbiguityFiltersDict() {
        return null;
    }

    protected abstract String getExtractType();

    protected NumberOptions getOptions() {
        return NumberOptions.None;
    }

    protected Optional<Pattern> getNegativeNumberTermsRegex() {
        return Optional.empty();
    }

    public List<ExtractResult> extract(String source) {

        if (source == null || source.isEmpty()) {
            return Collections.emptyList();
        }

        ArrayList<ExtractResult> result = new ArrayList<>();

        Boolean[] matched = new Boolean[source.length()];
        Arrays.fill(matched, false);

        HashMap<Match, String> matchSource = new HashMap<>();

        getRegexes().forEach((k, value) -> {

            Match[] matches = RegExpUtility.getMatches(k, source);

            for (Match m : matches) {
                int start = m.index;
                int length = m.length;
                for (int j = 0; j < length; j++) {
                    matched[start + j] = true;
                }

                // Keep Source Data for extra information
                matchSource.put(m, value);
            }
        });

        int last = -1;
        for (int i = 0; i < source.length(); i++) {

            if (matched[i]) {

                if (i + 1 == source.length() || !matched[i + 1]) {

                    int start = last + 1;
                    int length = i - last;
                    String subStr = source.substring(start, start + length);

                    int finalStart = start;
                    int finalLength = length;

                    Optional<Match> srcMatches = matchSource.keySet().stream().filter(o -> o.index == finalStart && o.length == finalLength).findFirst();

                    if (srcMatches.isPresent()) {
                        Match srcMatch = srcMatches.get();

                        // Extract negative numbers
                        if (getNegativeNumberTermsRegex().isPresent()) {

                            Matcher match = getNegativeNumberTermsRegex().get().matcher(source.substring(0, start));
                            if (match.find()) {
                                start = match.start();
                                length = length + (match.end() - match.start());
                                subStr = match.group() + subStr;
                            }
                        }

                        ExtractResult er = new ExtractResult(
                                start,
                                length,
                                subStr,
                                getExtractType(),
                                matchSource.containsKey(srcMatch) ? matchSource.get(srcMatch) : null);

                        result.add(er);
                    }
                }
            } else {
                last = i;
            }
        }

        result = filterAmbiguity(result, source);
        
        return result;
    }

    private ArrayList<ExtractResult> filterAmbiguity(ArrayList<ExtractResult> extractResults, String input) {
        if (getAmbiguityFiltersDict() != null) {
            for (Map.Entry<Pattern, Pattern> pair : getAmbiguityFiltersDict().entrySet()) {
                final Pattern key = pair.getKey();
                final Pattern value = pair.getValue();

                for (ExtractResult extractResult : extractResults) {
                    Optional<Match> keyMatch = Arrays.stream(RegExpUtility.getMatches(key, extractResult.getText())).findFirst();
                    if (keyMatch.isPresent()) {
                        final Match[] matches = RegExpUtility.getMatches(value, input);
                        extractResults = extractResults.stream()
                            .filter(er -> Arrays.stream(matches).noneMatch(m -> m.index < er.getStart() + er.getLength() && m.index + m.length > er.getStart()))
                            .collect(Collectors.toCollection(ArrayList::new));
                    }
                }
            }
        }

        return extractResults;
    }

    protected Pattern generateLongFormatNumberRegexes(LongFormatType type) {
        return generateLongFormatNumberRegexes(type, BaseNumbers.PlaceHolderDefault);
    }

    protected Pattern generateLongFormatNumberRegexes(LongFormatType type, String placeholder) {

        String thousandsMark = Pattern.quote(String.valueOf(type.thousandsMark));
        String decimalsMark = Pattern.quote(String.valueOf(type.decimalsMark));

        String regexDefinition = type.decimalsMark == '\0' ?
                BaseNumbers.IntegerRegexDefinition(placeholder, thousandsMark) :
                BaseNumbers.DoubleRegexDefinition(placeholder, thousandsMark, decimalsMark);

        return RegExpUtility.getSafeLookbehindRegExp(regexDefinition, Pattern.UNICODE_CHARACTER_CLASS);
    }
}
